### PR TITLE
update twig version and use namespaced twig classes

### PR DESCRIPTION
**Pull request by azjezz**
Adds PSR-4 Twig environment class alias.

allow users to access the twig environment using the PSR-4 twig environment class alias.

example : 
```php

use Twig\Environment;

$twig = $container->get(Environment::class);
```

- [x] Are you creating a new feature?
  - [x] Why is the new feature needed? What purpose does it serve?
  - [x] How will users use the new feature?
  - [x] Base your feature on the `develop` branch, and submit against that branch.
  - [x] Add only one feature per pull request; split multiple features over multiple pull requests

### CODE REVIEW

**Review by weierophinney — CHANGES_REQUESTED** [review]
I'd argue if you're going to do this, it should be done _for every Twig class we use_. Additionally, we should document using the namespaced versions, as this will make code written today forwards compatible with a future release of Twig that drops the legacy pseudo-namespaces.

One question I have: does our minimum supported version of Twig have the namespaced variants? If not, this is likely premature, and should accompany a change in minimum supported version. I'd also argue that if we do that, we should use the namespaced versions everywhere, and instead alias the pseudo-namespaced versions to the actual namespaced versions.

**Comment by geerteltink** [comment]
@weierophinney I've thought about this today and I think it's only needed for `Twig\Environment`. No other class is referenced for API use. The only extra requirement that is needed is to increase the minimum Twig version to `^1.34.4  || ^2.1`. For the rest I don't see any issues. 

In the next minor release we could even swap to PSR-4 classes usage internally without affecting current installations if they use ^1.34.

**Comment by azjezz** [comment]
@weierophinney i have update the minimum required twig version to `^1.34 | ^2.4` as 1.34 and 2.4 are the first twig versions to contain the namespaced class names

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
+.idea
+.idea/*
```
Comment (weierophinney) [inline .gitignore]: Please put these in your global gitignore file instead.

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,7 +10,7 @@ All notable changes to this project will be documented in this file, in reverse
 
 ### Changed
 
-- Nothing.
+-  update twig version and use namespaced twig classes #51.
```
Comment (weierophinney) [inline CHANGELOG.md]: Please use the same format as is used elsewhere in the changelog.

File: src/TwigEnvironmentFactory.php [hunk 3]
```diff
@@ -74,9 +75,12 @@
...
 {
     /**
-     * @throws Exception\InvalidConfigException for invalid config service values.
+     * @param ContainerInterface $container
+     *
+     * @return Environment
+     * @throws LoaderError
      */
-    public function __invoke(ContainerInterface $container) : TwigEnvironment
+    public function __invoke(ContainerInterface $container) : Environment
```
Comment (weierophinney) [inline src/TwigEnvironmentFactory.php:77]: For BC purposes, this should use the non-namespaced version in the signature. While the classes are equivalent, the namespaced version extends the non-namespaced one, so we need to keep the older one in the signature as it satisfies either return type.

It only needs to be reverted in this method, however; the other methods are private, so they do not affect the public API.

File: src/TwigExtension.php [hunk 4]
```diff
@@ -87,17 +88,24 @@ public function getFunctions() : array
...
-     *       RouteResult parameters will be used, defaults to true
+     * @param null|string $route
+     * @param array       $routeParams
+     * @param array       $queryParams
+     * @param null|string $fragmentIdentifier
+     * @param array       $options Can have the following keys:
+     *                             - reuse_result_params (bool): indicates if the current
+     *                             RouteResult parameters will be used, defaults to true
+     *
+     * @return string
```
Comment (weierophinney) [inline src/TwigExtension.php]: Please revert the changes to this docblock.

We only document parameters and return values that:

- do not have typehints, and/or
- have additional narrative descriptions.

Only `$options` falls under that category here.

File: src/TwigExtension.php [hunk 5]
```diff
@@ -110,17 +118,24 @@ public function renderUri(
...
```
Comment (weierophinney) [inline src/TwigExtension.php]: Please revert the changes to this docblock.

We only document parameters and return values that:

- do not have typehints, and/or
- have additional narrative descriptions.

Only `$options` falls under that category here.

File: src/TwigExtension.php [hunk 6]
```diff
@@ -131,6 +146,10 @@ public function renderUrl(
      *
      * Usage: {{ absolute_url('path/to/something') }}
      * Generates: http://example.com/path/to/something
+     *
+     * @param string|null $path
+     *
+     * @return string
```
Comment (weierophinney) [inline src/TwigExtension.php]: Please revert the changes to this docblock.

We only document parameters and return values that:

- do not have typehints, and/or
- have additional narrative descriptions.

Neither `$path` or the return value fall in those categories.

File: src/TwigExtension.php [hunk 7]
```diff
@@ -142,6 +161,11 @@ public function renderUrlFromPath(string $path = null) : string
      *
      * Usage: {{ asset('path/to/asset/name.ext', version=3) }}
      * Generates: path/to/asset/name.ext?v=3
+     *
+     * @param string      $path
+     * @param string|null $version
+     *
+     * @return string
```
Comment (weierophinney) [inline src/TwigExtension.php]: Please revert the changes to this docblock.

We only document parameters and return values that:

- do not have typehints, and/or
- have additional narrative descriptions.

None of these fall in those categories.

File: src/TwigRenderer.php [hunk 8]
```diff
@@ -36,16 +39,16 @@ class TwigRenderer implements TemplateRendererInterface
...
     protected $twigLoader;
 
     /**
-     * @var TwigEnvironment
+     * @var Environment
      */
     protected $template;
 
-    public function __construct(TwigEnvironment $template = null, string $suffix = 'html')
+    public function __construct(Environment $template = null, string $suffix = 'html')
```
Comment (weierophinney) [inline src/TwigRenderer.php:51]: We'll need to revert this to the non-namespaced version, for reasons listed in the factory earlier. That also applies to the `$template` and `$twigLoader` properties, as they are not private.

File: src/TwigRenderer.php [hunk 9]
```diff
@@ -65,25 +68,34 @@ public function __construct(TwigEnvironment $template = null, string $suffix = '
...
      * Render
      *
+     * @param string       $name
      * @param array|object $params
-     * @throws \Zend\Expressive\Template\Exception\InvalidArgumentException for non-array, non-object parameters.
+     *
+     * @return string
+     * @throws LoaderError
+     * @throws RuntimeError
+     * @throws SyntaxError
```
Comment (weierophinney) [inline src/TwigRenderer.php]: Please revert the `@param` and `@return` annotations, and _expand_ the `@throws` annotations to indicate what scenarios will result in those specific error types.

File: src/TwigRenderer.php [hunk 10]
```diff
@@ -100,10 +112,15 @@ public function render(string $name, $params = []) : string
 
     /**
      * Add a path for template
+     *
+     * @param string      $path
+     * @param string|null $namespace
+     *
```
Comment (weierophinney) [inline src/TwigRenderer.php]: Please revert the `@param` annotations.

File: src/TwigRenderer.php [hunk 11]
```diff
@@ -130,6 +147,10 @@ public function getPaths() : array
      *
      * Normalizes templates in the format "namespace::template" to
      * "@namespace/template".
+     *
+     * @param string $template
+     *
+     * @return string
```
Comment (weierophinney) [inline src/TwigRenderer.php]: Please revert the `@param` and `@return` annotations.

File: src/TwigRendererFactory.php [hunk 12]
```diff
@@ -29,7 +30,10 @@
 class TwigRendererFactory
 {
     /**
-     * @throws Exception\InvalidConfigException for invalid config service values.
+     * @param ContainerInterface $container
+     *
+     * @return TwigRenderer
+     * @throws LoaderError
```
Comment (weierophinney) [inline src/TwigRendererFactory.php]: Please revert the `@param` and `@return` annotations, and detail when a `LoaderError` is thrown.

File: src/TwigRendererFactory.php [hunk 13]
```diff
@@ -48,6 +52,9 @@ public function __invoke(ContainerInterface $container) : TwigRenderer
      * array having precedence.
      *
      * @param array|ArrayObject $config
+     *
+     * @return array
+     *
```
Comment (weierophinney) [inline src/TwigRendererFactory.php]: Throughout this class, revert any new `@param` or `@return` annotations for values that are already typehinted and have no additional descriptions. Add descriptions for any `@throws` annotations to indicate when those scenarios occur.

File: test/ConfigProviderTest.php [hunk 14]
```diff
@@ -34,6 +34,8 @@ public function testInvocationReturnsArray() : array
 
     /**
      * @depends testInvocationReturnsArray
+     *
+     * @param array $config
```
Comment (weierophinney) [inline test/ConfigProviderTest.php]: Revert, please.

File: test/ExceptionTest.php [hunk 15]
```diff
@@ -41,6 +41,8 @@ public function exception() : Generator
 
     /**
      * @dataProvider exception
+     *
+     * @param string $exception
```
Comment (weierophinney) [inline test/ExceptionTest.php]: Revert, please.

File: test/TestAsset/Extension/BarTwigExtension.php [hunk 16]
```diff
@@ -9,9 +9,11 @@
 
 namespace ZendTest\Expressive\Twig\TestAsset\Extension;
 
-class BarTwigExtension extends \Twig_Extension
+use Twig\Extension\AbstractExtension;
+
+class BarTwigExtension extends AbstractExtension
 {
-    public function getName()
+    public function getName(): string
```
Comment (weierophinney) [inline test/TestAsset/Extension/BarTwigExtension.php]: Our CS currently has spaces on either side of a `:` when used to denote a typehint.

File: test/TestAsset/Extension/FooTwigExtension.php [hunk 17]
```diff
@@ -9,9 +9,11 @@
 
 namespace ZendTest\Expressive\Twig\TestAsset\Extension;
 
-class FooTwigExtension extends \Twig_Extension
+use Twig\Extension\AbstractExtension;
+
+class FooTwigExtension extends AbstractExtension
 {
-    public function getName()
+    public function getName(): string
```
Comment (weierophinney) [inline test/TestAsset/Extension/FooTwigExtension.php]: Our CS currently has spaces on either side of a `:` when used to denote a typehint.

**Review by weierophinney — CHANGES_REQUESTED** [review]
I've flagged a few additional changes. First, there are a couple of potential BC breaks I need you to address; they're basically as easy as reverting the return type or parameter typehint to the non-namespaced variant in a few isolated cases. Second, it looks like your IDE went and populated a ton of `@param` and `@return` annotations that we frankly do not need (as they are redundant with the typehints already present).

File: src/TwigEnvironmentFactory.php (hunk 3 above)
Comment (azjezz) [reply src/TwigEnvironmentFactory.php:77]: actually, the [namespaced version](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php) doesn't extend the [non-namespaced one](https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php), its an alias . 
in the end of each non-namespaced class, [you can see that](https://github.com/twigphp/Twig/blob/2.x/lib/Twig/Environment.php#L970) and in the namespaced version class, it [imports the non-namespaced one](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php#L5) that's why all classes are within an [unreachable statement ](https://github.com/twigphp/Twig/blob/2.x/src/Environment.php#L7) so this [shouldn't cause any BC issues](https://3v4l.org/IW01f).

Comment (weierophinney) [reply src/TwigEnvironmentFactory.php:77]: I misread the first link (that's the file I'd found on which I'd based the argument that extension was occurring); the "extension" happens in a conditional that can never be triggered, while the alias happens instead in the class file defining `Twig_Environment` (I'm curious why they went this path, instead of doing the alias in the PSR-4 namespaced class file...).

That answers the BC questions for me, then. However, there are still the other issues to resolve.

**Comment by azjezz** [comment]
@weierophinney all set.

File: CHANGELOG.md (hunk 2 above)
Comment (azjezz) [reply CHANGELOG.md]: resolved.

File: src/TwigExtension.php (hunk 5 above)
Comment (azjezz) [reply src/TwigExtension.php]: resolved.

File: test/ConfigProviderTest.php (hunk 14 above)
Comment (azjezz) [reply test/ConfigProviderTest.php]: resolved.

File: test/TestAsset/Extension/BarTwigExtension.php (hunk 16 above)
Comment (azjezz) [reply test/TestAsset/Extension/BarTwigExtension.php]: resolved.

File: test/TestAsset/Extension/FooTwigExtension.php (hunk 17 above)
Comment (azjezz) [reply test/TestAsset/Extension/FooTwigExtension.php]: resolved.

File: test/ExceptionTest.php (hunk 15 above)
Comment (azjezz) [reply test/ExceptionTest.php]: resolved.

File: src/TwigExtension.php (hunk 4 above)
Comment (azjezz) [reply src/TwigExtension.php]: done :+1:

File: src/TwigRendererFactory.php [hunk 18]
```diff
@@ -28,9 +29,7 @@
  */
 class TwigRendererFactory
 {
-    /**
-     * @throws Exception\InvalidConfigException for invalid config service values.
-     */
+
```
Comment (weierophinney) [inline src/TwigRendererFactory.php:32]: Please revert - this method still throws that exception, via the `getEnvironment()` call.

File: src/TwigRendererFactory.php [hunk 19]
```diff
@@ -48,6 +47,7 @@ public function __invoke(ContainerInterface $container) : TwigRenderer
      * array having precedence.
      *
      * @param array|ArrayObject $config
+     *
```
Comment (weierophinney) [inline src/TwigRendererFactory.php:50]: Revert; we do not add whitespace between `@param`, `@return`, or `@throws` annotations.

File: test/TwigExtensionFunctionsRenderTest.php [hunk 20]
```diff
@@ -38,13 +38,14 @@ protected function setUp()
     /**
      * @param string $assetsUrl
      * @param string $assetsVersion
-     * @return Twig_Environment
+     *
+     * @return Environment
```
Comment (weierophinney) [inline test/TwigExtensionFunctionsRenderTest.php:42]: No whitespace between `@param` and `@return` annotations, please.

**Review by weierophinney — COMMENTED** [review]
A few more changes are needed, but, as they are CS/consistency issues, I'll make the changes during merge. Additionally, I'll be rebasing this against develop, as it _does_ indicate an improvement and change in functionality. I'll release this with the 2.2.0 release.

**Comment by weierophinney** [comment]
I've done the following:

- Rebased against develop
- Squashed commits to remove items from the history that are irrelevant (particularly the `.idea/` tree changes)
- Reworded the `CHANGELOG` entry and split it into two distinct changes
- Performed the CS/consistency fixes as noted above

All changes are now in the develop branch as of bad0be6
